Add table summarizing sampler properties and processing (#871)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

The OpenTelemetry API has two properties responsible for the data collection:

> * `IsRecording` field of a `Span`. If `true` the current `Span` records tracing

Reviewer: ```suggestion
* `IsRecording` field of a `Span`. If `true` the current `Span` records in memory tracing
```

Reviewer 2: What is "in memory data"? I think the current wording (== the old wording) is more clear here. What would be more clear IMHO would be:

```suggestion
* `IsRecording` field of a `Span`. If `false` the current `Span` discards all tracing
```

>   events (attributes, events, status, etc.), otherwise all tracing events are

Reviewer: ```suggestion
  data (attributes, events, status, etc.), otherwise all tracing data are
```

>   dropped. Users can use this property to determine if expensive trace events

Reviewer: ```suggestion
  dropped. Users can use this property to determine if expensive trace data
```

>   can be avoided. [Span Processors](#span-processor) will receive only those

Reviewer 2: ```suggestion
  can be avoided. [Span Processors](#span-processor) MUST receive only those
```

>   spans which has this field set to `true`. However, [Span

Reviewer: ```suggestion
  spans which have this field set to `true`. However, [Span
```

>   field set to `true`. However, [Span Exporter](#span-exporter) SHOULD not

Reviewer 2: Nit:
```suggestion
  field set to `true`. However, [Span Exporter](#span-exporter) SHOULD NOT
```

>   Exporter](#span-exporter) will not receive them unless the `Sampled` flag was

Reviewer 2: ```suggestion
  Exporter](#span-exporter) SHOULD not receive them unless the `Sampled` flag was
```

  also set.
* `Sampled` flag in `TraceFlags` on `SpanContext`. This flag is propagated via
  the `SpanContext` to child Spans. For more details see the [W3C Trace Context
  specification][trace-flags]. This flag indicates that the `Span` has been

>   `sampled` and will be exported. [Span Exporters](#span-exporter) will receive

Reviewer: ```suggestion
  `sampled` and will be passed to the [Span Exporters](#span-exporter), which will usually export it out of the process.
```

>   `sampled` and will be passed to the [Span Exporters](#span-exporter), which will usually export it out of the process.

Reviewer 2: > which will usually export it out of the process.

I don't think we should describe exporters here, this is about sampling.

>   `sampled` and will be exported. [Span Exporters](#span-exporter) will receive
>   only those spans with the `Sampled` flag set.

Reviewer 2: ```suggestion
  `sampled` and will be exported. [Span Exporters](#span-exporter) MUST receive
  those spans with the `Sampled` flag set (they SHOULD NOT receive the ones that don't).
```

>   receive those spans which have `Sampled` flag set to true and they SHOULD NOT receive the ones

Author: Since the text here is "SHOULD NOT" - it means the recommendation is to not call Exporters for Spans without `Sampled` flag set. Processors could still do it, and spec is not prohibiting it.
As we used  "SHOULD NOT" here, do we still require additional explicit clarification here, or in the table that "Depending on the Processor, Exporter may still get span without `Sampled` flag set? @carlosalberto @Oberon00

Reviewer 2: I don't think it's necessary but it would be nice. If you want, you can add a sentence to the built in span processors section that they MUST NOT call exporters for unsampled spans (unless explicitly requested).

Author: Let me address that as a follow up PR. I fear adding it might delay this PRs progress.

(Meanwhile I'll to figure out under what scenarios a user want to sent unsampled spans to exporter).

could cause gaps in the distributed trace, and because of this OpenTelemetry API
MUST NOT allow this combination.

The following table summarizes the expected behavior for each combination of
`IsRecording` and `SampledFlag`.

> | IsRecording | Sampled Flag | Span Processor receives Span? | Span Exporter receives Span? |

Reviewer 2: As @carlosalberto mentioned, the last column is up to the span processor (at least if it receives the span at all). So maybe change the heading to

```suggestion
| IsRecording | Sampled Flag | Span Processor receives Span? | Span Exporter receives Span (by default)? |
```

> | `IsRecording` | `Sampled` Flag | Span Processor receives Span? | Span Exporter receives Span? |

Reviewer 2: Actually the third column is a change / additional specification, but I think it's good.

@cijothomas Can you add "Fixes #782" to the PR description?

CC @alolita (you have #782 assigned currently).

| ------------- | -------------- | ----------------------------- | ---------------------------- |
| true          | true           | true                          | true                         |
| true          | false          | true                          | false                        |

> | false         | true           | Not allowed                   | Not allowed                  |

Reviewer 3: What's the difference between `Not allowed` and `false`?

Author: Not allowed means this combination is disallowed by the spec. False means processor is not called.
